### ISSUE TITLE
Getting None from Downcast

### ISSUE THREAD

**Issue by aztecrex** [fogti/as-any#3]
I can't get this implementation to work despite it being fundementally the same as an implementation I wrote.

I had developed an AsAny trait inside a project I was working on. After I got it working I searched for a library that implemented something similar and I found this project. Your trait and implementation are exactly the same as the one I had developed with the exceptions:
* mine has `as_any` and `as_any_mut` only, not `type_name` 
* mine has no `inline` directives
* mine doesn't have the blanket implementation (what I was about to do when I started looking for libs)

The implementation of the two traits is exactly the same.

Yet, when I use the local version, everything is fine. When I drop in the `AsAny` from this library my `downcast_ref` calls return `None`

Not even sure how to debug something like this.

**Comment by aztecrex** [comment]
Actually, NVM. It's because of the blanket implementation. In my case, the blanket imp causes `as_any` to return a reference to an outer box instead of the value in the box.

**Comment by zseri** [comment]
yeah, autoderef and such can make it very non-obvious what's going on. I usually use explicit reference types when I want to control that more strictly...

**Comment by ssrlive** [comment]
I meet the same issue. here it my code, but `obj.mmmm()` never called. how to resolve it?
```rust
trait MyTrait: std::fmt::Debug + AsAny {
    fn my_method(&self);
}

#[derive(Debug)]
struct MyStruct1;
impl MyTrait for MyStruct1 {
    fn my_method(&self) {
        println!("MyStruct1 my_method");
    }
}

impl MyStruct1 {
    fn mmmm(&self) {
        println!("MyStruct1 mmmm");
    }
}

#[derive(Debug)]
struct MyStruct2;
impl MyTrait for MyStruct2 {
    fn my_method(&self) {
        println!("MyStruct2 my_method");
    }
}

fn main() {
        let mut my_vec: Vec<Box<dyn MyTrait>> = Vec::new();
        my_vec.push(Box::new(MyStruct1));
        my_vec.push(Box::new(MyStruct2));

        for item in &my_vec {
            item.my_method();
            println!("item.type_name = {}", item.type_name());

            item.downcast_ref::<MyStruct1>().map(|obj| obj.mmmm());
        }
}
```

**Comment by fogti** [comment]
If someone has suggestions about how to document this potential pitfall, please open a PR to insert that into the library docs...
